python/cudf/cudf/core/reshape.py

```diff
@@ -12,6 +12,30 @@
 _axis_map = {0: 0, 1: 1, "index": 0, "columns": 1}
 
 
+def _align_objs(objs, how="outer"):
+    """Align a set of Series or Dataframe objects.
+
+    Parameters
+    ----------
+    objs : list of DataFrame, Series, or Index
+
+    Returns
+    -------
+    A set of reindexed and aligned objects ready for concatenation
+    """
+    index = objs[0].index
+    for obj in objs[1:]:
+        name = index.name
+        index = (
+            cudf.DataFrame(index=obj.index)
+            .join(cudf.DataFrame(index=index), how=how)
+            .index
+        )
+        index.name = name
+
+    return [obj.reindex(index) for obj in objs]
```

Comment:
Still need to handle the case where all of the `obj.index` are equals for this.

Reply:
That check is done at the top of concat when `axis=`.

```
    if axis == 1:
        # Check if multiindex then check if indexes match. GenericIndex
        # returns ndarray tuple of bools requiring additional filter.
        # Then check for duplicate index value.
        i_objs = iter(objs)
        first = next(i_objs)
        if all(isinstance(o.index, cudf.MultiIndex) for o in objs):
            match_index = all(first.index == rest.index for rest in i_objs)
        else:
            match_index = all(
                all(first.index == rest.index) for rest in i_objs
            )

        if not match_index:
            if not all(o.index.is_unique for o in objs):
                raise ValueError("cannot reindex from a duplicate axis")
```

Reply:
Apologies, missed that! Would like to move that into the `_align_objs` function as the behavior is consistent across Pandas when aligning indexes.